### PR TITLE
Implement the concept of Environment Record

### PR DESCRIPTION
**Pull request by ericmorand-sonarsource**
Fixes JS-175

### CODE REVIEW

File: packages/jsts/src/dbd/js-to-dbd/core/expressions/literal.ts [hunk 1]
```diff
@@ -1,21 +1,45 @@
...
+  }
+
+  let value = valueByConstantTypeRegistry.get(typeof constant.value);
+
+  if (!value) {
+    value = createValue(typeof constant.value, context.scopeManager.createValueIdentifier());
+
+    valueByConstantTypeRegistry.set(typeof constant.value, value);
+
+    instructions.push(
```
Comment (zglicz) [inline packages/jsts/src/dbd/js-to-dbd/core/expressions/literal.ts:33]: This should be pushed immediately to the current block

**Review by zglicz — APPROVED** [review]
Once I actually started having a grasp on the code, you go and re-write it all.

I left 1 comment, as I think you missed a direct push instead of adding to an instructions array.

File: packages/jsts/src/dbd/js-to-dbd/core/expressions/literal.ts (hunk 1 above)
Comment (ericmorand-sonarsource) [reply packages/jsts/src/dbd/js-to-dbd/core/expressions/literal.ts:33]: Nice catch!

**Comment by ericmorand-sonarsource** [comment]
> Once I actually started having a grasp on the code, you go and re-write it all.

That's a fair point, and I'm very sorry about that. When trying to make the first sample pass, I ended up writing so many conditionals that I had the feeling something was wrong. I really hope that this is the last major change that we have to make. Please accept my apologies, we should have made those changes together.

**Pull request merged**